spec/src/main/asciidoc/converters.asciidoc

```diff
@@ -104,3 +104,19 @@ If no built-in nor custom `Converter` exists for a requested Type `T`, an implic
 
 If a `Converter` implements the `java.lang.AutoCloseable` interface  then the `close()` method will be called when the underlying `Config` is being released.
 
+=== Empty value
+
+A config property can be configured with an empty value (the empty String `""`).
+Such config property will be interpreted by the Config API as follows:
+
+* `Config.getValue(key, String.class)` returns `""` (empty String)
```

Comment:
I disagree strongly with this behavior.  This makes it impossible to have a configuration source which overrides a present value with a non-present value.

Reply:
yes, this is against the possibility to detect 'nils'. This would break other features, like to 'unconfigure' values.

Reply:
The spec never mentions anything about "unconfiguring" the properties.
My understanding of the spec is that the configsource only allows to overrides values, not to "delete/unconfigure" them.

If a lower configsource defines `a=foo` and I override that with a higher configsource to be `a= ` (empty value). I expect the Config API to take my value and returns ``.

Reply:
Probably we never wrote this down but I only explained this in the past. The original handing in DeltaSpike was to have an empty string 'erase' any configured value in a lower ordinal ConfigSource. This turned out to be quite useful. Especially now with `Optional<T>` it is otherwise possible to come back to the orElse. Do you have a better idea?

Reply:
> The spec never mentions anything about "unconfiguring" the properties.
> My understanding of the spec is that the configsource only allows to overrides values, not to "delete/unconfigure" them.

This to me is a mistake and we should amend the specification.

> If a lower configsource defines `a=foo` and I override that with a higher configsource to be `a= ` (empty value). I expect the Config API to take my value and returns ``.

As I stated elsewhere, there is no possible meaningful distinction between empty and missing.  At *best* it encourages poor behavior and UX.

Reply:
I agree with @dmlloyd proposal to modify the Optional case:


Input string | Output type | Current rules | Better rules
-- | -- | -- | --
"" | Optional<String> | Optional.of("") | Optional.empty()

This would make it possible to use the `Optional<T>` to "unconfigure" the value and use the `orElse`.

However, I'm not convinced and need to think more how changing the behaviour for the plain `getValue`.

If I understand you correctly, If I have a lower config source with `a=foo` that I want to configure, I'd have to provide a higher config source with `a=`.

In my code, I then have to take into account all the use cases:
* `a` is not configured anywhere, `Config.getValue(String)` throws `NoSuchElementException`
* `a` is configured in a config source, `Config.getValue(String)` return `foo`
* `a` is configured in a config source and "unconfigured" in a higher config source with an empty string. What does `Config.getValue()` return?
  * `` (empty string that I specified)
  *  `null` as David is proposing
  *  throws a `NoSuchElementException` (current rule and it is consistent with "unconfiguring" the property)